pandas/core/reshape.py

```diff
@@ -1257,19 +1275,41 @@ def get_empty_Frame(data, sparse):
         sp_indices = [[] for _ in range(len(dummy_cols))]
         for ndx, code in enumerate(codes):
             if code == -1:
-                # Blank entries if not dummy_na and code == -1, #GH4446
-                continue
-            sp_indices[code].append(ndx)
+                if dummy_na != 'keep':
+                    # Blank entries if not dummy_na and code == -1, #GH4446
+                    continue
+                else:
+                    # Use a negative index here to code NaNs if
+                    # we are keeping them.
+                    # Offset by -1 to account for zero.
+                    # Have to add to ALL columns.
+                    for _code in range(len(dummy_cols)):
+                        sp_indices[_code].append(-ndx - 1)
+            else:
+                sp_indices[code].append(ndx)
 
         if drop_first:
             # remove first categorical level to avoid perfect collinearity
             # GH12042
             sp_indices = sp_indices[1:]
             dummy_cols = dummy_cols[1:]
         for col, ixs in zip(dummy_cols, sp_indices):
-            sarr = SparseArray(np.ones(len(ixs), dtype=np.uint8),
-                               sparse_index=IntIndex(N, ixs), fill_value=0,
-                               dtype=np.uint8)
+            if dummy_na == 'keep':
+                sarr = np.ones(len(ixs), dtype=np.float32)
+                # NaNs are marked by a negative index.
+                ixs = np.array(ixs)
+                sarr[ixs < 0] = np.nan
+                ixs[ixs < 0] += 1  # undo the offset
+                ixs = np.abs(ixs)  # set index back to positive.
+                sarr = SparseArray(
+                    sarr,
+                    sparse_index=IntIndex(N, ixs), fill_value=0,
+                    dtype=np.float32)
+            else:
+                sarr = SparseArray(
```

Comment:
can you pull the SparseArray creation out of the if/then. you can simply pass in an already created ``sarr`` (of the correctly dtype; don't actually need to pass the dtype.